### PR TITLE
PPS geometry for Run3

### PR DESCRIPTION
**Pull request by fabferro**
#### PR description:
This PR implements the description of Run3 PPS detectors (tracking and timing) at the status of art of current knowledge (2021/Run3 geometry for PPS was till now a replica of Run2 geometry).
Two geometries have been implemented: one to be used in Simulation workflows and one in Reconstruction's.
This is a peculiarity of PPS that we need to conserve and that arises from the specific features of how we deal with the connection with LHC machine in the two workflows.
Most of the files can be used for both geometries, when it was not possible, the involved files have been placed in Reco/Simu subdirs.  
The geometry described in dict2021Geometry.py P5 contains the Simu version, the one described in geometryRPFromDD_2021_cfi.py contains the Reco version.
We have placed the new files in directories marked with 2021 flags trying to comply as much as possible to the Geometry naming conventions, provided that we also had to deal with two different detectors placed in four different pots, and a strict observations of the rules was not considered efficient nor clear enough.  
Files needed in Run2 geometries have not been touched, to keep a full back-compatibility.
The tools to produce .db files for 2021 geometry has been update3d as well.

#### PR validation:
The geometries have been tested with a full (private) Simu+Reco chain which loads the two geometries, checking that we obtain compatible results.
Same results are obtained reading directly the xml's and reading the .db files produced with the official tools.
Tests with Run2 data has also been done: Run2 data processing is not affected by the PR.
The detectors are still being built thus no tests on real data could be performed.

### CODE REVIEW

**Comment by fabferro** [comment]
FYI @wpcarvalho

**Comment by civanch** [comment]
please test

**Comment by cvuosalo** [comment]
@fabferro I don't understand the title of this PR. pre4 is done. It was made last week. This PR might be able to get into pre5, but pre4 is past.

File: Geometry/VeryForwardGeometry/python/dd4hep/v2 [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (cvuosalo) [inline Geometry/VeryForwardGeometry/python/dd4hep/v2:1]: This file does not have a correct name.

**Comment by cvuosalo** [comment]
`Configuration/Geometry/python/generate2021Geometry.py` needs to be run, and the files it changes need to be included in this PR. Also `Configuration/Geometry/README.md` needs to be updated.

**Comment by fabferro** [comment]
> 
> 
> `Configuration/Geometry/python/generate2021Geometry.py` needs to be run, and the files it changes need to be included in this PR. Also `Configuration/Geometry/README.md` needs to be updated.

With Python generate2021Geometry.py -P5 I could generate only the files I added (like Geometry/CMSCommonData/data/cmsExtendedGeometry2021XML_cfi.py). Can you tell me the syntax to generate all the other ones and to modify the meaning of the standard flags (I mean: 2021= T3+C1+M1+P4+O3+F1, etc.)?
Thanks.

**Comment by fabferro** [comment]
> 
> 
> @fabferro I don't understand the title of this PR. pre4 is done. It was made last week. This PR might be able to get into pre5, but pre4 is past.

In a hurry I left an automatic title. I changed it. Sorry.

File: Geometry/VeryForwardGeometry/python/dd4hep/v2 (hunk 1 above)
Comment (fabferro) [reply Geometry/VeryForwardGeometry/python/dd4hep/v2]: It should be a dir actually. Something went wrong... I fixed it.

**Comment by kpedro88** [comment]
@fabferro 
> 
> 
> > `Configuration/Geometry/python/generate2021Geometry.py` needs to be run, and the files it changes need to be included in this PR. Also `Configuration/Geometry/README.md` needs to be updated.
> 
> With Python generate2021Geometry.py -P5 I could generate only the files I added (like Geometry/CMSCommonData/data/cmsExtendedGeometry2021XML_cfi.py). Can you tell me the syntax to generate all the other ones and to modify the meaning of the standard flags (I mean: 2021= T3+C1+M1+P4+O3+F1, etc.)?
> Thanks.

These are set in https://github.com/cms-sw/cmssw/blob/fce3630957ba32ef548d6156ce46f8306e27beed/Configuration/Geometry/python/dict2021Geometry.py#L2358-L2365.

**Comment by civanch** [comment]
@cvuosalo , @kpedro88 , is it correct, that PPS developers should apply Configuration/Geometry/python/generate2021Geometry.py? I would expect, that the new geometry files are tested privately committed to the repository and after your overview you make a simple PR, which will include new developments.

**Comment by civanch** [comment]
We are mixing ~50 new files for PPS and configuration of full geometry for CMS.

**Comment by kpedro88** [comment]
All Run3 and Phase2 detector geometry configurations are centrally managed.

Whether the P5 version of PPS should be included in the default 2021 configuration, or a separate configuration should be made, should be decided by those planning the Run3 campaigns.

**Comment by civanch** [comment]
So, we speak only on creation of P5 in this PR.

**Comment by kpedro88** [comment]
If a new top-level detector geometry configuration is outside the scope of this PR, then any changes to the centrally-managed 2021 geometry configuration files should be reverted.

File: Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Reco/v1/RP_Cuts_Per_Region.xml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+<DDDefinition xmlns="http://www.cern.ch/cms/DDL"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../DDLSchema/DDLSchema.xsd">
+  <SpecParSection label="RP_Cuts_Per_Region.xml" eval="true">
+    <SpecPar name="RP_Silicon_Detectors_Range_Cuts">
+      <PartSelector path="//RP_Silicon_Detector"/>
+      <Parameter name="CMSCutsRegion" value="RPSiliconDetectorsCuts" eval="false"/>
+      <!--Parameter name="ProdCutsForElectrons" value="0.10*mm"/>
+      <Parameter name="ProdCutsForPositrons" value="0.10*mm"/>
+      <Parameter name="ProdCutsForGamma" value="0.10*mm"/-->
```
Comment (civanch) [inline Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Reco/v1/RP_Cuts_Per_Region.xml:11]: @fabferro , we have here duplicated lines for cuts.

File: Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Simu/v1/CTPPS_Cuts_Per_Region.xml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+<DDDefinition xmlns="http://www.cern.ch/cms/DDL"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../DDLSchema/DDLSchema.xsd">
+  <SpecParSection label="RP_Cuts_Per_Region.xml" eval="true">
+    <SpecPar name="RP_Silicon_Detectors_Range_Cuts">
+      <PartSelector path="//RP_Silicon_Detector"/>
+      <Parameter name="CMSCutsRegion" value="RPSiliconDetectorsCuts" eval="false"/>
+      <!--Parameter name="ProdCutsForElectrons" value="0.10*mm"/>
+      <Parameter name="ProdCutsForPositrons" value="0.10*mm"/>
+      <Parameter name="ProdCutsForGamma" value="0.10*mm"/-->
```
Comment (civanch) [inline Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Simu/v1/CTPPS_Cuts_Per_Region.xml:11]: here again positron and gamma cuts are repeated twice.

File: Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Reco/v1/RP_Cuts_Per_Region.xml (hunk 2 above)
Comment (fabferro) [reply Geometry/VeryForwardData/data/CTPPS_2021/Cuts_Per_Region/Reco/v1/RP_Cuts_Per_Region.xml:11]: The lines are commented, but I can remove them anyway.

**Comment by fabferro** [comment]
Reverted the common files to the original version as discussed in the meeting. P5 section left in dict2021 but not (yet) used as 2021 baseline.

**Comment by perrotta** [comment]
please test

**Comment by fabferro** [comment]
Investigating test2021Geometry failure:
===== Test "test2021Geometry" ====
cmsExtendedGeometry2021ZeroMaterialXML_cfi.py differs
cmsExtendedGeometry2021ZeroMaterial.xml differs
cmsExtendedGeometry2021FlatPlus10PercentXML_cfi.py differs
cmsExtendedGeometry2021FlatPlus10Percent.xml differs
cmsExtendedGeometry2021FlatMinus10PercentXML_cfi.py differs
cmsExtendedGeometry2021FlatMinus10Percent.xml differs
cmsExtendedGeometry2021FlatPlus05PercentXML_cfi.py differs
cmsExtendedGeometry2021FlatPlus05Percent.xml differs
cmsExtendedGeometry2021FlatMinus05PercentXML_cfi.py differs
cmsExtendedGeometry2021FlatMinus05Percent.xml differs
cmsExtendedGeometry2021XML_cfi.py differs
cmsExtendedGeometry2021.xml differs

---> test test2021Geometry had ERRORS
TestTime:1
^^^^ End Test test2021Geometry ^^^^

They do differ because I did not generate them, as decided at the Simulation meeting. I reverted in dict2021Geometry file the standard to P4, instead of P5. I hope it can fix this failure. 
@cvuosalo @civanch

**Comment by fabferro** [comment]
As to "test TestDQMServicesDemo had ERRORS", ther's a SEG FAULT I have no clue of the reason.
Does it depend on this PR?

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@fabferro ,  I have requested tests. If will fail again, let us ask help of @cvuosalo .

**Comment by qliphy** [comment]
> As to "test TestDQMServicesDemo had ERRORS", ther's a SEG FAULT I have no clue of the reason.
> Does it depend on this PR?
It is related to https://github.com/cms-sw/cmssw/issues/34520
and should be fixed by https://github.com/cms-sw/cmssw/pull/34532

**Comment by cvuosalo** [comment]
+1

**Comment by fabferro** [comment]
Today is last open day, I hope this PR can go in 12_0_0.

**Comment by ggovi** [comment]
+1

**Comment by cvuosalo** [comment]
urgent

**Comment by perrotta** [comment]
@cms-sw/upgrade-l2 please have a look

**Comment by civanch** [comment]
@srimanob, please, review this PR - it seems to be OK now.

**Comment by srimanob** [comment]
@fabferro 
Sorry, could you please mention what agrees in the SIM meeting? I am not so sure. The P5 you add in this PR does not go to any detector version as far as I see. Is this recommendation from the meeting?

Note that, 12_0_0_pre5 is postponed to next week. However, I think this PR can merge after clarification. 

Big thanks, and sorry for delay.

**Comment by fabferro** [comment]
> 
> 
> @fabferro
> Sorry, could you please mention what agrees in the SIM meeting? I am not so sure. The P5 you add in this PR does not go to any detector version as far as I see. Is this recommendation from the meeting?
> 
> Note that, 12_0_0_pre5 is postponed to next week. However, I think this PR can merge after clarification.
> 
> Big thanks, and sorry for delay.

At the Simulation meeting ot was agreed that P5 will become the deafault version of the PPS geometry as soon as the Simulation code itself will be up to date and ready to be dealt by means of centralized processing. It was also agreed that, if self consistent, the files added by this PR could be pulled in CMSSW. 
As PPS Offline coordinator, I think it would be a big step forward for us to have this PR merged already for 12_0_0.

**Comment by srimanob** [comment]
+Upgrade

This PR is to add a new geometry for PPS. The new geometry is currently off until the new updated simulation code will be ready for the later PR.

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
@fabferro 
We are getting many IB errors after this PR getting merged. For example:
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/pyRelValMatrixLogs/run/11650.911_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA/step1_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA.log#/40-40
edm::FileInPath unable to find file Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml anywhere in the search path.

Please make a PR to upload those xml files to the repository the sooner the better:
https://github.com/cms-data/Geometry-VeryForwardData

e.g. the above xml should be put under a directory RP_Vertical_Device/2021/v1 in the repository

**Comment by fabferro** [comment]
> @fabferro
> We are getting many IB errors after this PR getting merged. For example:
> https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/pyRelValMatrixLogs/run/11650.911_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA/step1_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA.log#/40-40
> edm::FileInPath unable to find file Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml anywhere in the search path.
> 
> Please make a PR to upload those xml files to the repository the sooner the better:
> https://github.com/cms-data/Geometry-VeryForwardData
> 
> e.g. the above xml should be put under a directory RP_Vertical_Device/2021/v1 in the repository

Working on it...

**Comment by smuzaffar** [comment]
Currently all the xml files are in cmssw/Geometry/VeryForwardData . If there is not strong reason to keep these files in cmssw then pls move cmssw/Geometry/VeryForwardData/data to https://github.com/cms-data/Geometry-VeryForwardData

**Comment by qliphy** [comment]
Moreover, @fabferro  as you moved
[1] Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
to
[2] Geometry/VeryForwardData/data/RP_Vertical_Device/2021/Simu/v1/RP_Vertical_Device.xml

this brings errors to complain
edm::FileInPath unable to find file Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml anywhere in the search path.
in  e.g.
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/unitTestLogs/DetectorDescription/DDCMS#/547-547
and 
===== Test "cmsRun /data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_12_0_X_2021-07-21-2300/src/Geometry/TrackerGeometryBuilder/test/python/testTrackerHierarchy_cfg.py " ====

Should we keep [1] instead?

**Comment by fabferro** [comment]
> > @fabferro
> > We are getting many IB errors after this PR getting merged. For example:
> > https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/pyRelValMatrixLogs/run/11650.911_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA/step1_ZMM_14+2021_DD4hep+ZMM_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA.log#/40-40
> > edm::FileInPath unable to find file Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml anywhere in the search path.
> > Please make a PR to upload those xml files to the repository the sooner the better:
> > https://github.com/cms-data/Geometry-VeryForwardData
> > e.g. the above xml should be put under a directory RP_Vertical_Device/2021/v1 in the repository
> 
> Working on it...

It seems I miss the permission to write to the repository.

**Comment by qliphy** [comment]
> @qliphy It seems I miss the permission to write to the repository.

@fabferro You can make a PR similar as to cms-sw

**Comment by fabferro** [comment]
> Moreover, @fabferro as you moved
> [1] Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
> to
> [2] Geometry/VeryForwardData/data/RP_Vertical_Device/2021/Simu/v1/RP_Vertical_Device.xml
> 
> this brings errors to complain:
> edm::FileInPath unable to find file Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml anywhere in the search path.
> in e.g.
> https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/unitTestLogs/DetectorDescription/DDCMS#/547-547
> and
> https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-21-2300/unitTestLogs/Geometry/TrackerGeometryBuilder#/73-73
> 
> Should we keep [1] instead?

No, we should have both. One is used for P4 and one for P5. The version used for P4 can't be deleted until P5 becomes the default and P4 deprecated.

**Comment by qliphy** [comment]
> No, we should have both. One is used for P4 and one for P5. The version used for P4 can't be deleted until P5 becomes the default and P4 deprecated.

Yes, I know.   My question is that you have already moved
Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
to 
Geometry/VeryForwardData/data/RP_Vertical_Device/2021/Simu/v1/RP_Vertical_Device.xml

but there are other codes still requiring 
Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
so should we keep it? or should we touch those other codes?

**Comment by fabferro** [comment]
> > No, we should have both. One is used for P4 and one for P5. The version used for P4 can't be deleted until P5 becomes the default and P4 deprecated.
> 
> Yes, I know. My question is that you have already moved
> Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
> to
> Geometry/VeryForwardData/data/RP_Vertical_Device/2021/Simu/v1/RP_Vertical_Device.xml
> 
> but there are other codes still requiring
> Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml
> so should we keep it? or should we touch those other codes?

We should keep both even if they are the same file. We moved it but we should have copied it. We are adding Run3 geometry, reordering Geometry/VeryForwardData/data and keeping a full back compatibility. This PR is an intermediate step.

**Comment by qliphy** [comment]
Thanks @fabferro   I saw you now add the missing file in https://github.com/cms-data/Geometry-VeryForwardData/pull/1

However, as suggested by @smuzaffar  https://github.com/cms-sw/cmssw/pull/34499#issuecomment-884746696
it would be better to move all xml files under cmssw/Geometry/VeryForwardData/data to https://github.com/cms-data/Geometry-VeryForwardData
Would you please do that?

**Comment by fabferro** [comment]
> Thanks @fabferro I saw you now add the missing file in [cms-data/Geometry-VeryForwardData#1](https://github.com/cms-data/Geometry-VeryForwardData/pull/1)
> 
> However, as suggested by @smuzaffar [#34499 (comment)](https://github.com/cms-sw/cmssw/pull/34499#issuecomment-884746696)
> it would be better to move all xml files under cmssw/Geometry/VeryForwardData/data to https://github.com/cms-data/Geometry-VeryForwardData
> Would you please do that?

I can copy them but this sounds new to me. Are you going to ask me to remove the files from CMSSW?
@cvuosalo can you comment on this request?

**Comment by qliphy** [comment]
@fabferro  
1) You can first make a cms-data PR to copy all xml files under cmssw/Geometry/VeryForwardData/data to https://github.com/cms-data/Geometry-VeryForwardData
2) then you can make a cms-sw PR to remove all those xml files
3) We will take care of afterwards steps

On the other hand, we can also first merge https://github.com/cms-data/Geometry-VeryForwardData/pull/1 first to add back the missing file,  in order to cure the IB issues.  Then you can take actions (above (1) and (2) ) later.

**Comment by fabferro** [comment]
> @fabferro
> 
>     1. You can first make a cms-data PR to copy all xml files under cmssw/Geometry/VeryForwardData/data to https://github.com/cms-data/Geometry-VeryForwardData
> 
>     2. then you can make a cms-sw PR to remove all those xml files
> 
>     3. We will take care of afterwards steps
> 
> 
> On the other hand, if there is no objection, we can also first merge [cms-data/Geometry-VeryForwardData#1](https://github.com/cms-data/Geometry-VeryForwardData/pull/1) to add back the missing file, in order to cure the IB issues. Then you can take actions (above (1) and (2) ) later.

Let's see if with the PR to cms-data we have cured the IB issues.
Before proceeding to move all PPS xml files to cms-data I would like to understand it better. 
1- Are all detectors moving the files to cms-data? I still see tons of them in CMSSW.
2- Will I be able to link them directly in .py or in .xml files inside CMSSW?
As I said it's the first time I hear such a request.
Thanks.

**Comment by qliphy** [comment]
> Before proceeding to move all PPS xml files to cms-data I would like to understand it better.
> 1- Are all detectors moving the files to cms-data? I still see tons of them in CMSSW.

It is now recommended to move files to cms-data.  

> 2- Will I be able to link them directly in .py or in .xml files inside CMSSW?

Yes, it works in many other cases.

**Comment by cvuosalo** [comment]
I must have missed an important announcement. When has it been discussed to move Geometry XML files out of CMSSW?

**Comment by cvuosalo** [comment]
@fabferro I think `Geometry/VeryForwardData/data/RP_Vertical_Device/2021/v1/RP_Vertical_Device.xml` should be restored in CMSSW.

**Comment by qliphy** [comment]
> I must have missed an important announcement. When has it been discussed to move Geometry XML files out of CMSSW?

@smuzaffar or @silviodonato may comment more on this.  But I notice this is now more recommended at least for new data files.

As we receive complains about missing file holding Run3 test,  @fabferro would you please make a PR to cms-sw to add the missing file?   We can then discuss about the migration of geometry XML files to cms-data or not at e.g. next ORP.

**Comment by fabferro** [comment]
> > I must have missed an important announcement. When has it been discussed to move Geometry XML files out of CMSSW?
> 
> @smuzaffar or @silviodonato may comment more on this. But I notice this is now more recommended at least for new data files.
> 
> As we receive complains about missing file holding Run3 test, @fabferro would you please make a PR to cms-sw to add the missing file? We can then discuss about the migration of geometry XML files to cms-data or not at e.g. next ORP.

Sure. Doing it.

**Comment by fabferro** [comment]
Done in PR#34588. Code checks ongoing.

**Pull request merged**